File: simpsonApproximation.py
#Author: Kyle Mayer
#Simpson Approximation

'''
Process:
    -Get interval from lower and upper bounds
    -Use the interval to calculate midpoint interval
    -Calculate summation for left and right ends of rectangles
    -Use left and right end calculations to get trapezoids.
        Just average left and right.
    -Calculate summation of midpoint rectangles    
    -Take weighted average of trapezoids and midpoint rectangles. (2T+M)/3

    Result of Simpson approximation is an close approximation of an indefinite
    integral of some function. 


    Notes: 
    - I still need to define a neater way to switch functions. Very
    clunky and needs to be changed in multiple places as of this version.
    - Add ability for user to input bounds and desired equation?
    - Use matplotlib to show visual of area of indefinite integral?
    - GUI?
'''

import math

#define bounds
lowerBound = 1.0
upperBound = 3.0

#changing numIntervals increases precision
numIntervals = 1000000000

#find interval (b-a)/2
interval = (upperBound - lowerBound) / float(numIntervals)

#midpoint interval will always be half of interval used for L and R
mInterval = interval / 2.0 

#set starting points for each distributed summation
lStart = lowerBound
rStart = lowerBound + interval
mStart = lowerBound + mInterval

#init variables for each result
l = 0.0
r = 0.0
t = 0.0
m = 0.0
s = 0.0

#### Calculations
# Current function: 1/x

for i in range(0, numIntervals): 
    #dealing with L
    l += interval * (math.sqrt(lStart))
    #increment by interval
    lStart += interval
    
    #dealing with R
    if(rStart <= upperBound):
        r += interval * (math.sqrt(rStart))
        #increment by interval
        rStart += interval

#get average of L and R
t = (l + r) / 2

for i in range(0, numIntervals):
    #dealing with M
    if(mStart <= upperBound):
        m += interval * (math.sqrt(mStart))
        mStart += interval

#finally get simpson approximation using weighted average
s = ((2 * m) + t) / 3

print("         ----Results----")
print("Left end rectangle:  " + str(l))
print("Right end rectangle: " + str(r))
print("Trapezoid:           " + str(t)) 
print("Midpoint:            " + str(m))
print("Simpson:             " + str(s))

File: simpsonApproximation.cpp
//Author: Kyle Mayer
//Simpson Approximation

/*
 * To compile:
 *      g++ simpsonApproximation.cpp -o simpsonApproximation
 *
 * To run:
 *      ./simpsonApproximation
 *
 *
 *  Notes:
 *      - See Python file for notes on what could be done to improve. 
*/


#include <iostream>
#include <math.h>

using namespace std;

int main() 
{
    double lowerBound = 1.0;
    double upperBound = 3.0;

    int numIntervals = 1000000000;

    double interval = (upperBound - lowerBound) / static_cast<double>(numIntervals);

    double mInterval = interval / 2.0;

    double lStart = lowerBound;
    double rStart = lowerBound + interval;
    double mStart = lowerBound + mInterval;

    double l = 0.0;
    double r = 0.0;
    double t = 0.0;
    double m = 0.0;
    double s = 0.0;

    //calculations

    for(int i = 0; i < numIntervals; i++)
    {
        //dealing with L
        l += interval * sqrt(lStart);
        lStart += interval;

        //dealing with L
        if(rStart <= upperBound)
        {
            r += interval * sqrt(rStart);
            rStart += interval;
        }
    }    

    //get average of L and R
    t = (l + r) / 2.0;

    for(int i = 0; i < numIntervals; i++)
    {
        if(mStart <= upperBound)
        {
            m += interval * sqrt(mStart);
            mStart += interval;
        }
    }

    s = ((2 * m) + t) / 3.0;

    cout.precision(20);
    cout << "\n         -----Results-----\n";
    cout << "Left end rectangle:     " << l << "\n";
    cout << "Right end rectangle:    " << r << "\n";
    cout << "Trapezoid:              " << t << "\n";
    cout << "Midpoint:               " << m << "\n";
    cout << "Simpson:                " << s << "\n";

}
